src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -446,18 +470,23 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
                 lengthForConstructor = elementCount;
             }
 
-            if (!inPlace && !used)
+            if (inPlaceTarget is null && !used)
             {
                 // emitting a value that no one will see
                 return true;
             }
 
             if (elementCount == 0)
             {
-                if (inPlace)
+                if (inPlaceTarget is not null)
                 {
+                    EmitAddress(inPlaceTarget, Binder.AddressKind.Writeable);
                     _builder.EmitOpCode(ILOpCode.Initobj);
                     EmitSymbolToken(spanType, wrappedExpression.Syntax);
+                    if (used)
+                    {
+                        EmitExpression(inPlaceTarget, used: true);
```

Comment:
>EmitExpression(inPlaceTarget, used: true); [](http://example.com/codeflow?start=24&length=42)

It is not obvious what is going on here. Are we potentially going to evaluate an arbitrary complex expression twice?

Reply:
I see that ```inPlaceTarget``` is going to be either a local or parameter. Consider adding ```Debug.Assert(inPlaceTarget is null || TargetIsNotOnHeap(inPlaceTarget), "in-place construction target should not be on heap");``` at the beginning of the function.

Reply:
Done